File: CALQULUS-PMS/src/mobile/google/OfflineSyncManager.kt
package com.calqulusrms.app

import android.content.Context
import androidx.room.Room
import androidx.room.RoomDatabase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withContext

class OfflineSyncManager private constructor(private val context: Context) {
    private val database: CALQULUS RMSDatabase = Room.databaseBuilder(
        context,
        CALQULUS RMSDatabase::class.java,
        "calqulusrms.db"
    ).build()
    
    private val _syncStatus = MutableStateFlow<SyncStatus>(SyncStatus.Idle)
    val syncStatus: StateFlow<SyncStatus> = _syncStatus
    
    private val _lastSyncDate = MutableStateFlow<Date?>(null)
    val lastSyncDate: StateFlow<Date?> = _lastSyncDate
    
    private val _pendingChanges = MutableStateFlow(0)
    val pendingChanges: StateFlow<Int> = _pendingChanges
    
    private val _isOnline = MutableStateFlow(true)
    val isOnline: StateFlow<Boolean> = _isOnline
    
    companion object {
        @Volatile
        private var instance: OfflineSyncManager? = null
        
        fun getInstance(context: Context): OfflineSyncManager {
            return instance ?: synchronized(this) {
                instance ?: OfflineSyncManager(context).also { instance = it }
            }
        }
    }
    
    suspend fun sync() {
        if (!_isOnline.value) {
            _syncStatus.value = SyncStatus.Offline
            return
        }
        
        _syncStatus.value = SyncStatus.Syncing
        
        try {
            // Upload pending changes
            uploadPendingChanges()
            
            // Download latest data
            downloadLatestData()
            
            // Update sync status
            _syncStatus.value = SyncStatus.Completed
            _lastSyncDate.value = Date()
            _pendingChanges.value = 0
            
        } catch (e: Exception) {
            _syncStatus.value = SyncStatus.Failed
            println("Sync error: ${e.message}")
        }
    }
    
    private suspend fun uploadPendingChanges() {
        val pendingChanges = database.pendingChangeDao().getPendingChanges()
        
        for (change in pendingChanges) {
            APIClient.uploadChange(change)
            
            // Mark as synced
            change.isSynced = true
            database.pendingChangeDao().update(change)
        }
    }
    
    private suspend fun downloadLatestData() {
        // Download properties
        val properties = APIClient.getProperties()
        
        // Download tenants
        val tenants = APIClient.getTenants()
        
        // Download maintenance requests
        val maintenanceRequests = APIClient.getMaintenanceRequests()
        
        // Save to local database
        saveProperties(properties)
        saveTenants(tenants)
        saveMaintenanceRequests(maintenanceRequests)
    }
    
    private suspend fun saveProperties(properties: List<Property>) {
        val existingProperties = database.propertyDao().getAll()
        
        for (property in properties) {
            val existing = existingProperties.find { it.id == property.id }
            
            if (existing != null) {
                // Update existing
                existing.name = property.name
                existing.address = property.address
                existing.units = property.units
                database.propertyDao().update(existing)
            } else {
                // Create new
                database.propertyDao().insert(property)
            }
        }
    }
    
    private suspend fun saveTenants(tenants: List<Tenant>) {
        val existingTenants = database.tenantDao().getAll()
        
        for (tenant in tenants) {
            val existing = existingTenants.find { it.id == tenant.id }
            
            if (existing != null) {
                // Update existing
                existing.name = tenant.name
                existing.email = tenant.email
                existing.unitId = tenant.unitId
                database.tenantDao().update(existing)
            } else {
                // Create new
                database.tenantDao().insert(tenant)
            }
        }
    }
    
    private suspend fun saveMaintenanceRequests(requests: List<MaintenanceRequest>) {
        val existingRequests = database.maintenanceRequestDao().getAll()
        
        for (request in requests) {
            val existing = existingRequests.find { it.id == request.id }
            
            if (existing != null) {
                // Update existing
                existing.title = request.title
                existing.description = request.description
                existing.status = request.status
                database.maintenanceRequestDao().update(existing)
            } else {
                // Create new
                database.maintenanceRequestDao().insert(request)
            }
        }
    }
    
    suspend fun addPendingChange(type: String, data: String) {
        val change = PendingChange(
            id = java.util.UUID.randomUUID().toString(),
            type = type,
            data = data,
            isSynced = false,
            createdAt = System.currentTimeMillis()
        )
        
        database.pendingChangeDao().insert(change)
        _pendingChanges.value = database.pendingChangeDao().getPendingChangesCount()
    }
    
    suspend fun getLocalProperties(): List<Property> {
        return database.propertyDao().getAll()
    }
    
    suspend fun getLocalTenants(): List<Tenant> {
        return database.tenantDao().getAll()
    }
    
    suspend fun getLocalMaintenanceRequests(): List<MaintenanceRequest> {
        return database.maintenanceRequestDao().getAll()
    }
    
    fun setOnlineStatus(online: Boolean) {
        _isOnline.value = online
    }
}

enum class SyncStatus {
    Idle,
    Syncing,
    Completed,
    Failed,
    Offline
}

// Room Database
@androidx.room.Database(
    entities = [
        Property::class,
        Tenant::class,
        MaintenanceRequest::class,
        PendingChange::class
    ],
    version = 1
)
abstract class CALQULUS RMSDatabase : RoomDatabase() {
    abstract fun propertyDao(): PropertyDao
    abstract fun tenantDao(): TenantDao
    abstract fun maintenanceRequestDao(): MaintenanceRequestDao
    abstract fun pendingChangeDao(): PendingChangeDao
}

// Entities
@androidx.room.Entity(tableName = "properties")
data class Property(
    @androidx.room.PrimaryKey val id: String,
    val name: String,
    val address: String,
    val units: Int
)

@androidx.room.Entity(tableName = "tenants")
data class Tenant(
    @androidx.room.PrimaryKey val id: String,
    val name: String,
    val email: String,
    val unitId: String
)

@androidx.room.Entity(tableName = "maintenance_requests")
data class MaintenanceRequest(
    @androidx.room.PrimaryKey val id: String,
    val title: String,
    val description: String,
    val status: String
)

@androidx.room.Entity(tableName = "pending_changes")
data class PendingChange(
    @androidx.room.PrimaryKey val id: String,
    val type: String,
    val data: String,
    var isSynced: Boolean = false,
    val createdAt: Long
)

// DAOs
@androidx.room.Dao
interface PropertyDao {
    @androidx.room.Query("SELECT * FROM properties")
    suspend fun getAll(): List<Property>
    
    @androidx.room.Insert
    suspend fun insert(property: Property)
    
    @androidx.room.Update
    suspend fun update(property: Property)
    
    @androidx.room.Delete
    suspend fun delete(property: Property)
}

@androidx.room.Dao
interface TenantDao {
    @androidx.room.Query("SELECT * FROM tenants")
    suspend fun getAll(): List<Tenant>
    
    @androidx.room.Insert
    suspend fun insert(tenant: Tenant)
    
    @androidx.room.Update
    suspend fun update(tenant: Tenant)
    
    @androidx.room.Delete
    suspend fun delete(tenant: Tenant)
}

@androidx.room.Dao
interface MaintenanceRequestDao {
    @androidx.room.Query("SELECT * FROM maintenance_requests")
    suspend fun getAll(): List<MaintenanceRequest>
    
    @androidx.room.Insert
    suspend fun insert(request: MaintenanceRequest)
    
    @androidx.room.Update
    suspend fun update(request: MaintenanceRequest)
    
    @androidx.room.Delete
    suspend fun delete(request: MaintenanceRequest)
}

@androidx.room.Dao
interface PendingChangeDao {
    @androidx.room.Query("SELECT * FROM pending_changes WHERE isSynced = 0")
    suspend fun getPendingChanges(): List<PendingChange>
    
    @androidx.room.Query("SELECT COUNT(*) FROM pending_changes WHERE isSynced = 0")
    suspend fun getPendingChangesCount(): Int
    
    @androidx.room.Insert
    suspend fun insert(change: PendingChange)
    
    @androidx.room.Update
    suspend fun update(change: PendingChange)
    
    @androidx.room.Delete
    suspend fun delete(change: PendingChange)
}

File: CALQULUS-PMS/src/mobile/apple/OfflineSyncManager.swift
import SwiftUI
import CoreData
import Combine

class OfflineSyncManager: ObservableObject {
    static let shared = OfflineSyncManager()
    
    @Published var syncStatus: SyncStatus = .idle
    @Published var lastSyncDate: Date?
    @Published var pendingChanges: Int = 0
    @Published var isOnline: Bool = true
    
    private let persistentContainer: NSPersistentContainer
    private var cancellables = Set<AnyCancellable>()
    
    private init() {
        persistentContainer = NSPersistentContainer(name: "CalqulusRMS")
        persistentContainer.loadPersistentStores { description, error in
            if let error = error {
                print("Core Data error: \(error)")
            }
        }
        
        setupNetworkMonitoring()
    }
    
    var viewContext: NSManagedObjectContext {
        return persistentContainer.viewContext
    }
    
    private func setupNetworkMonitoring() {
        // Monitor network connectivity
        // In production, this would use NWPathMonitor
        isOnline = true
    }
    
    func sync() async {
        guard isOnline else {
            syncStatus = .offline
            return
        }
        
        syncStatus = .syncing
        
        do {
            // Upload pending changes
            try await uploadPendingChanges()
            
            // Download latest data
            try await downloadLatestData()
            
            // Update sync status
            syncStatus = .completed
            lastSyncDate = Date()
            pendingChanges = 0
            
        } catch {
            syncStatus = .failed
            print("Sync error: \(error)")
        }
    }
    
    private func uploadPendingChanges() async throws {
        let pendingChanges = fetchPendingChanges()
        
        for change in pendingChanges {
            try await APIClient.shared.uploadChange(change)
            
            // Mark as synced
            change.isSynced = true
        }
        
        try viewContext.save()
    }
    
    private func downloadLatestData() async throws {
        // Download properties
        let properties = try await APIClient.shared.getProperties()
        
        // Download tenants
        let tenants = try await APIClient.shared.getTenants()
        
        // Download maintenance requests
        let maintenanceRequests = try await APIClient.shared.getMaintenanceRequests()
        
        // Save to local database
        try saveProperties(properties)
        try saveTenants(tenants)
        try saveMaintenanceRequests(maintenanceRequests)
    }
    
    private func fetchPendingChanges() -> [NSManagedObject] {
        let request = NSFetchRequest<NSManagedObject>(entityName: "PendingChange")
        request.predicate = NSPredicate(format: "isSynced == NO")
        
        return try? viewContext.fetch(request) ?? []
    }
    
    private func saveProperties(_ properties: [Property]) throws {
        let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: "Property")
        let existingProperties = try? viewContext.fetch(fetchRequest)
        
        for property in properties {
            if let existing = existingProperties?.first(where: { ($0.value(forKey: "id") as? String) == property.id }) {
                // Update existing
                existing.setValue(property.name, forKey: "name")
                existing.setValue(property.address, forKey: "address")
                existing.setValue(property.units, forKey: "units")
            } else {
                // Create new
                let entity = NSEntityDescription.insertNewObject(forEntityName: "Property", into: viewContext)
                entity.setValue(property.id, forKey: "id")
                entity.setValue(property.name, forKey: "name")
                entity.setValue(property.address, forKey: "address")
                entity.setValue(property.units, forKey: "units")
            }
        }
        
        try viewContext.save()
    }
    
    private func saveTenants(_ tenants: [Tenant]) throws {
        let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: "Tenant")
        let existingTenants = try? viewContext.fetch(fetchRequest)
        
        for tenant in tenants {
            if let existing = existingTenants?.first(where: { ($0.value(forKey: "id") as? String) == tenant.id }) {
                // Update existing
                existing.setValue(tenant.name, forKey: "name")
                existing.setValue(tenant.email, forKey: "email")
                existing.setValue(tenant.unitId, forKey: "unitId")
            } else {
                // Create new
                let entity = NSEntityDescription.insertNewObject(forEntityName: "Tenant", into: viewContext)
                entity.setValue(tenant.id, forKey: "id")
                entity.setValue(tenant.name, forKey: "name")
                entity.setValue(tenant.email, forKey: "email")
                entity.setValue(tenant.unitId, forKey: "unitId")
            }
        }
        
        try viewContext.save()
    }
    
    private func saveMaintenanceRequests(_ requests: [MaintenanceRequest]) throws {
        let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: "MaintenanceRequest")
        let existingRequests = try? viewContext.fetch(fetchRequest)
        
        for request in requests {
            if let existing = existingRequests?.first(where: { ($0.value(forKey: "id") as? String) == request.id }) {
                // Update existing
                existing.setValue(request.title, forKey: "title")
                existing.setValue(request.description, forKey: "description")
                existing.setValue(request.status, forKey: "status")
            } else {
                // Create new
                let entity = NSEntityDescription.insertNewObject(forEntityName: "MaintenanceRequest", into: viewContext)
                entity.setValue(request.id, forKey: "id")
                entity.setValue(request.title, forKey: "title")
                entity.setValue(request.description, forKey: "description")
                entity.setValue(request.status, forKey: "status")
            }
        }
        
        try viewContext.save()
    }
    
    func addPendingChange(type: String, data: Data) {
        let entity = NSEntityDescription.insertNewObject(forEntityName: "PendingChange", into: viewContext)
        entity.setValue(UUID().uuidString, forKey: "id")
        entity.setValue(type, forKey: "type")
        entity.setValue(data, forKey: "data")
        entity.setValue(false, forKey: "isSynced")
        entity.setValue(Date(), forKey: "createdAt")
        
        do {
            try viewContext.save()
            pendingChanges += 1
        } catch {
            print("Error saving pending change: \(error)")
        }
    }
    
    func getLocalProperties() -> [Property] {
        let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: "Property")
        
        guard let entities = try? viewContext.fetch(fetchRequest) else {
            return []
        }
        
        return entities.compactMap { entity in
            guard let id = entity.value(forKey: "id") as? String,
                  let name = entity.value(forKey: "name") as? String,
                  let address = entity.value(forKey: "address") as? String,
                  let units = entity.value(forKey: "units") as? Int else {
                return nil
            }
            
            return Property(id: id, name: name, address: address, units: units)
        }
    }
    
    func getLocalTenants() -> [Tenant] {
        let fetchRequest = NSFetchRequest<NSManagedObject>(entityName: "Tenant")
        
        guard let entities = try? viewContext.fetch(fetchRequest) else {
            return []
        }
        
        return entities.compactMap { entity in
            guard let id = entity.value(forKey: "id") as? String,
                  let name = entity.value(forKey: "name") as? String,
                  let email = entity.value(forKey: "email") as? String,
                  let unitId = entity.value(forKey: "unitId") as? String else {
                return nil
            }
            
            return Tenant(id: id, name: name, email: email, unitId: unitId)
        }
    }
}

enum SyncStatus {
    case idle
    case syncing
    case completed
    case failed
    case offline
}

struct Property: Codable {
    let id: String
    let name: String
    let address: String
    let units: Int
}

struct Tenant: Codable {
    let id: String
    let name: String
    let email: String
    let unitId: String
}

struct MaintenanceRequest: Codable {
    let id: String
    let title: String
    let description: String
    let status: String
}
